Softeq.XToolkit.Common/Extensions/TaskExtensions.cs

```diff
@@ -11,43 +11,79 @@ namespace Softeq.XToolkit.Common.Extensions
-        /// <typeparam name="T">Task type.</typeparam>
-        public static async Task<T> WithTimeout<T>(this Task<T> task, TimeSpan timeout)
+        /// <param name="task">The task to wait for.</param>
+        /// <param name="timeout">The maximum time to wait.</param>
+        /// <returns>
+        ///     A task that completes with the result of the specified <paramref name="task"/> or
+        ///     faults with a <see cref="TimeoutException"/> if <paramref name="timeout"/> elapses first.
+        /// </returns>
+        public static async Task WithTimeout(this Task task, TimeSpan timeout)
         {
-            var retTask = await Task.WhenAny(task, Task.Delay((int) timeout.TotalMilliseconds))
-                .ConfigureAwait(false);
+            // Details:
+            //     https://devblogs.microsoft.com/pfxteam/crafting-a-task-timeoutafter-method/
+            // Source:
+            //     https://github.com/microsoft/vs-threading/blob/master/src/Microsoft.VisualStudio.Threading/TplExtensions.cs#L73
 
-            if (retTask is Task<T>)
+            if (task == null)
             {
-                return task.Result;
+                throw new ArgumentNullException(nameof(task));
             }
 
-            return default;
+            using (var timerCancellation = new CancellationTokenSource())
+            {
+                var timeoutTask = Task.Delay(timeout, timerCancellation.Token);
+                var firstCompletedTask = await Task.WhenAny(task, timeoutTask).ConfigureAwait(false);
+
+                if (firstCompletedTask == timeoutTask)
+                {
+                    throw new TimeoutException();
+                }
+
+                timerCancellation.Cancel();
+
+                await task.ConfigureAwait(false);
+            }
         }
 
-        public static Task<T> SafeTaskWrapper<T>(this Task<T> task, ILogger logger = null)
+        /// <summary>
+        ///     Returns a task that completes as the original task completes or when a timeout expires, whichever happens first.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by the original task.</typeparam>
+        /// <param name="task">The task to wait for.</param>
+        /// <param name="timeout">The maximum time to wait.</param>
+        /// <returns>
+        ///     A task that completes with the result of the specified <paramref name="task"/> or
+        ///     faults with a <see cref="TimeoutException"/> if <paramref name="timeout"/> elapses first.
+        /// </returns>
+        public static async Task<T> WithTimeout<T>(this Task<T> task, TimeSpan timeout)
         {
-            task.ContinueWith(t => { LogException(t, logger); }, CancellationToken.None,
-                TaskContinuationOptions.NotOnRanToCompletion, TaskScheduler.Default);
-            return task;
+            await WithTimeout((Task) task, timeout).ConfigureAwait(false);
+            return task.GetAwaiter().GetResult();
```

Comment:
Can this task just be awaited here?

Reply:
Not need, the task already has result.

Reply:
Exactly! That's why there's no difference in using await and .GetAwaiter().GetResult() - both will return result immediately. But .GetAwaiter().GetResult() is still a blocking synchronous code in async method, i'd say - avoid it at all costs

Reply:
Avoid it at all costs, yes. But depends on cases.
In the current case `await` has the same calls under the hood.
The problem can be in case when `WithTimeoutAsync` method will be changed.